File: Исходники/Глава 2. Часть 1/Жадные алгоритмы/9. Алгоритм Хаффмана/Python/Haffman.py
from queue import PriorityQueue
from collections import deque


class Edge:
    def __init__(self, fr0m, to):
        self.fr0m = fr0m
        self.to = to


class HaffmanTree:
    def __init__(self, letters: dict[str: int]):
        self._letters = letters
        self._tree = self._build_tree_haffman()
        if self._tree != -1:
            self._root = self._get_root(self._tree)
        else:
            self._root = -1

    @staticmethod
    def _get_root(tree):
        nodes = sorted(tree, key=lambda x: len(x), reverse=True)
        return nodes[0]

    @staticmethod
    def _tree_adjacency_list(tree_edges):
        adjacency_list = dict()
        for edge in tree_edges:
            adjacency_list[edge.fr0m] = list()
            adjacency_list[edge.to] = list()
        for edge in tree_edges:
            adjacency_list[edge.fr0m].append(edge.to)
        return adjacency_list

    def _build_tree_haffman(self):
        if len(letters) > 1:
            edges_tree = list()
            free_nodes = PriorityQueue()
            for data, frequency in self._letters.items():
                free_nodes.put((frequency, data))
            while free_nodes.qsize() > 1:
                left = free_nodes.get()
                right = free_nodes.get()
                parent = (left[0] + right[0], left[1] + right[1])
                free_nodes.put(parent)
                edges_tree.append(Edge(parent[1], left[1]))
                edges_tree.append(Edge(parent[1], right[1]))
            return self._tree_adjacency_list(edges_tree)
        return -1

    @staticmethod
    def _encoding(node, letter, tree):
        if len(node) > 1:
            if letter in tree[node][0]:
                return '0'
            elif letter in tree[node][1]:
                return '1'
        return str()

    def encode_processing(self, word):
        code, visited = str(), list()
        queue = deque()
        if self._root == -1:
            return
        for letter in word:
            queue.append(self._root)
            while queue:
                node = queue.popleft()
                if node not in visited:
                    code += self._encoding(node, letter, self._tree)
                    if node == letter:
                        queue.clear(), visited.clear()
                    else:
                        queue += self._tree[node]
                        visited.append(node)
        return code

    def decode_processing(self, row):
        count, result = 0, str()
        letter = self._root
        if letter == -1:
            return
        while count < len(row):
            digit = row[count]
            if digit == '0':
                letter = self._tree[letter][0]
            else:
                letter = self._tree[letter][1]
            if len(letter) == 1:
                result += letter
                letter = self._root
            count += 1
        return result

    def get_all_codes(self):
        codes = dict()
        if len(letters) > 1:
            for letter in self._letters:
                codes[letter] = self.encode_processing(letter)
            return codes


letters = {'a': 15, 'b': 7, "f": 23}
tree_haffman = HaffmanTree(letters)
print(tree_haffman.get_all_codes())
print(tree_haffman.encode_processing('abf'))
#print(tree_haffman.decode_processing('1101100101000010'))

File: Исходники/Глава 2. Часть 1/Жадные алгоритмы/9. Алгоритм Хаффмана/C++/Олимпиадный/Haffman.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <set>
#include <vector>
#include <queue>
#include <map>

using namespace std;


class Edge {
	public:
		string from;
		string to;

		Edge(string curr_from, string curr_to) {
			from = curr_from;
			to = curr_to;
		}
};


class HaffmanTree {
public:
	HaffmanTree(vector<pair<int, string>> dict_letters) {
		letters = dict_letters;
		tree = build_haffman_tree();
		root = get_root(tree);
	}

	string encode_processing(string word) {
		string code;
		set<string> visited;
		string start = root;
		deque<string> queue_1;
		for (char letter : word) {
			queue_1.push_back(start);
			while (!queue_1.empty()) {
				string node = queue_1.front();
				queue_1.pop_front();
				if (visited.count(node) == 0) {
					code += encoding(node, letter, tree);
					if (node.size() == 1 and node.find(letter) != string::npos) {
						visited.clear();
						queue_1.clear();
					}
					else {
						if (!tree[node].empty()){
							queue_1.push_back(tree[node][0]);
							queue_1.push_back(tree[node][1]);
						}
						visited.insert(node);
					}
				}
			}
		}
		return code;
	}

	string decode_processing(string row) {
		int count = 0;
		string result;
		string letter = root;
		while (count < row.size()) {
			char digit = row[count];
			if (digit == '0') {
				letter = tree[letter][0];
			}
			else {
				letter = tree[letter][1];
			}
			if (letter.size() == 1) {
				result += letter;
				letter = root;
			}
			count += 1;
		}
		return result;
	}

	map<string, string> get_all_codes() {
		map<string, string> codes;
		for (auto letter : letters) {
			codes[letter.second] = encode_processing(letter.second);
		}
		return codes;
	}

private:
	vector<pair<int, string>> letters;
	map<string, vector<string>> tree;
	string root;


	static bool compare_root(string node_1, string node_2) {
		return node_1.size() > node_2.size();
	}


	static string get_root(map<string, vector<string>> tree) {
		vector<string> keys;
		for (auto key : tree) {
			keys.push_back(key.first);
		}
		sort(keys.begin(), keys.end(), compare_root);
		return keys[0];
	}


	static map<string, vector<string>> tree_adjacency_list(vector<Edge> tree_edges) {
		map<string, vector<string>> adjacency_list;
		for (Edge edge : tree_edges) {
			adjacency_list[edge.from] = vector<string>{};
			adjacency_list[edge.to] = vector<string>{};
		}
		for (Edge edge : tree_edges) {
			adjacency_list[edge.from].push_back(edge.to);
		}
		return adjacency_list;
	}


	map<string, vector<string>> build_haffman_tree() {
		vector<Edge> edges_tree;
		priority_queue<pair<int, string>, 
			vector<pair<int, string>>, 
			greater<pair<int, string>>> free_nodes;
		for (pair<int, string> node : letters) {
			free_nodes.push(node);
		}
		while (free_nodes.size() > 1) {
			pair<int, string> node_1 = free_nodes.top();
			free_nodes.pop();
			pair<int, string> node_2 = free_nodes.top();
			free_nodes.pop();
			pair<int, string> parent = {node_1.first + node_2.first, node_1.second + node_2.second};
			free_nodes.push(parent);
			
			edges_tree.push_back(Edge(parent.second, node_1.second));
			edges_tree.push_back(Edge(parent.second, node_2.second));
		}
		return tree_adjacency_list(edges_tree);
	}


	static string encoding(string node, char letter, map<string, vector<string>> tree) {
		if (node.size() > 1) {
			if (tree[node][0].find(letter) != string::npos) {
				return "0";
			}
			else if (tree[node][1].find(letter) != string::npos) {
				return "1";
			}
		}
		return "";
	}
};

//baf a
//ba a
//f a
//b a
//a a
//baf b
//ba b
//f b
//b b
//baf f
//ba f
//f f

//
//int main() {
//	vector<pair<int, string>>  letters = { {15, "a"}, {7, "b"}, {23, "f"}};
//	HaffmanTree haffman_tree(letters);
//	cout << haffman_tree.encode_processing("abf") << endl;
//	//cout << haffman_tree.decode_processing("1101100101000010") << endl;
//	//for (auto i : haffman_tree.get_all_codes()) {
//	//	cout << i.first << " " << i.second << endl;
//	//}
//	return 0;
//}